File: framework/languages/cpp/e2e/PubSub/Client/main.cpp
/* SPDX-License-Identifier: MPL-2.0 */

#include "Scenarios/fanout_basic_delivery_scenario.hpp"
#include "Scenarios/late_subscriber_scenario.hpp"
#include "Scenarios/missing_message_name_scenario.hpp"
#include "Scenarios/publisher_restart_scenario.hpp"
#include "Scenarios/slow_subscriber_scenario.hpp"
#include "Scenarios/subscriber_reconnect_scenario.hpp"
#include "Scenarios/topic_filter_scenario.hpp"
#include "Support/client_support.hpp"

#include <exception>
#include <iostream>
#include <string>

namespace ps_client = zlink::framework::e2e::pubsub::client;

int main ()
{
    try {
        ps_client::touch_file (ps_client::env_or ("ZLINK_CPP_E2E_START_READY_FILE"));
        ps_client::wait_for_file (ps_client::env_or ("ZLINK_CPP_E2E_START_CONTINUE_FILE"));
        std::this_thread::sleep_for (std::chrono::milliseconds (1000));

        const auto scenario = ps_client::env_or ("ZLINK_CPP_E2E_SCENARIO", "basic");
        const auto publisher_url = ps_client::env_or ("ZLINK_CPP_E2E_PUBLISHER_URL");
        ps_client::ensure (!publisher_url.empty (), "ZLINK_CPP_E2E_PUBLISHER_URL is required");

        if (scenario == "basic") {
            ps_client::run_fanout_basic_delivery_scenario (publisher_url);
        } else if (scenario == "topic") {
            ps_client::run_topic_filter_scenario (publisher_url);
        } else if (scenario == "late") {
            ps_client::run_late_subscriber_scenario (publisher_url);
        } else if (scenario == "reconnect") {
            ps_client::run_subscriber_reconnect_scenario (publisher_url);
        } else if (scenario == "slow") {
            ps_client::run_slow_subscriber_scenario (publisher_url);
        } else if (scenario == "publisher-restart") {
            ps_client::run_publisher_restart_scenario (publisher_url);
        } else if (scenario == "negative") {
            ps_client::run_missing_message_name_scenario (publisher_url);
        } else {
            throw std::runtime_error ("unknown scenario " + scenario);
        }

        std::cout << "pubsub e2e result=passed\n";
        return 0;
    }
    catch (const std::exception &error) {
        std::cerr << "pubsub scenario failed: " << error.what () << "\n";
        return 1;
    }
}

File: framework/languages/node/e2e/PubSub/Client/main.ts
import { parseClientOptions } from './Support/client-options';
import { ServerProcessLauncher, type DynamicProcess } from './Support/server-process-launcher';
import { runPsA1 } from './Scenarios/ps-a1-fanout-basic-delivery-scenario';
import { runPsA2 } from './Scenarios/ps-a2-topic-filter-scenario';
import { runPsA3 } from './Scenarios/ps-a3-late-subscriber-scenario';
import { runPsA4 } from './Scenarios/ps-a4-subscriber-reconnect-scenario';
import { runPsB1 } from './Scenarios/ps-b1-slow-subscriber-scenario';
import { runPsB2 } from './Scenarios/ps-b2-publisher-restart-scenario';
import { runPsC1 } from './Scenarios/ps-c1-missing-message-name-scenario';

async function main(): Promise<void> {
  const options = parseClientOptions(process.argv.slice(2));
  const processes = new ServerProcessLauncher(options);
  let restartedPublisher: DynamicProcess | undefined;
  const subscribers = options.subscriberUrls;

  const scenarios: Record<string, () => Promise<void>> = {
    'PS-A1': () => runPsA1(options.publisherUrl, subscribers),
    'PS-A2': () => runPsA2(options.publisherUrl, subscribers),
    'PS-A3': () => runPsA3(options.publisherUrl, options.lateSubscriberUrl, processes, options.publisherEndpoint),
    'PS-A4': () => runPsA4(options.publisherUrl, options.lateSubscriberUrl, subscribers.slice(0, 2), processes, options.publisherEndpoint),
    'PS-B1': () => runPsB1(options.publisherUrl, subscribers.slice(0, 2), subscribers[subscribers.length - 1]),
    'PS-B2': async () => { restartedPublisher = await runPsB2(options.publisherUrl, subscribers, processes, options.publisherEndpoint); },
    'PS-C1': () => runPsC1(options.publisherUrl, subscribers)
  };

  try {
    if (options.scenario.toLowerCase() === 'all') {
      for (const scenario of Object.values(scenarios)) {
        await scenario();
      }
    } else {
      const selected = scenarios[options.scenario.toUpperCase()];
      if (selected === undefined) {
        throw new Error(`Unknown scenario '${options.scenario}'.`);
      }
      await selected();
    }
  } finally {
    if (restartedPublisher !== undefined && !restartedPublisher.hasExited) {
      await restartedPublisher.stop();
    }
  }

  console.log('pubsub e2e result=passed');
}

main().catch((error: unknown) => {
  console.error(error);
  process.exitCode = 1;
});
